google/datalab/contrib/pipeline/_pipeline.py

```diff
@@ -104,6 +92,40 @@ def py(self):
     return Pipeline._imports + default_args + dag_definition + \
         task_definitions + up_steam_statements
 
+  @staticmethod
+  def _get_default_args(email, start_date, end_date):
+    start_date_str = Pipeline._get_datetime_expr_str(start_date)
+    end_date_str = Pipeline._get_datetime_expr_str(end_date)
+    airflow_default_args_format = """
+default_args = {{
+    'owner': 'Datalab',
+    'depends_on_past': False,
+    'email': ['{0}'],
+    'start_date': {1},
+    'end_date': {2},
+    'email_on_failure': True,
+    'email_on_retry': False,
+    'retries': 1,
+    'retry_delay': timedelta(minutes=1),
+}}
+
+"""
+    return airflow_default_args_format.format(email, start_date_str, end_date_str)
+
+  @staticmethod
+  def _get_datetime_expr_str(datetime_obj):
+    # User is expected to always provide start_date and end_date in UTC and in
+    # the %Y-%m-%dT%H:%M:%SZ format (i.e. _with_ the trailing 'Z' to
+    # signify UTC).
+    # However, due to a bug/feature in yaml.load(), strings that look like
+    # datetimes are parsed into timezone _unaware_ datetime objects (even if
+    # they do have a timezone). Hence we will be highly restrictive in the
+    # format we accept and will only allow '%Y-%m-%dT%H:%M:%SZ' format.
```

Comment:
This comment says we only allow a format that ends with Z, but the format on the next line does not end with Z. Is one of these incorrect?

Reply:
tl;dr - both are correct. 

For now, '%Y-%m-%dT%H:%M:%SZ' will be our advertised format of accepting UTC date-times. 

The '%Y-%m-%dT%H:%M:%S' is an implementation detail. I'm using it to specify the datetime, and then am explicitly setting the timezone('UTC') on the datetime (L126).

Eventually, once I implement yaml validation against a schema (https://b.corp.google.com/issues/64951979#comment2), I'll throw an error when users do _not_ specify their string in the '%Y-%m-%dT%H:%M:%SZ' format.

Reply:
Perhaps you could add a TODO about what needs to be done and when.